cluster-autoscaler/simulator/cluster_snapshot.go

```diff
@@ -0,0 +1,56 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package simulator
+
+import (
+	apiv1 "k8s.io/api/core/v1"
+	schedulerlisters "k8s.io/kubernetes/pkg/scheduler/listers"
+)
+
+// ClusterSnapshot is abstraction of cluster state used for predicate simulations.
+// It exposes efficient mutation methods and can be viewed as scheduler's SharedLister
+// via GetSchedulerLister() method.
+type ClusterSnapshot interface {
+	// AddNode adds node to the snapshot.
+	AddNode(node *apiv1.Node) error
+	// RemoveNode removes nodes (and pods scheduled to it) from the snapshot.
+	RemoveNode(nodeName string) error
+	// AddPod adds pod to the snapshot and schedules it to given node.
+	AddPod(pod *apiv1.Pod, nodeName string) error
+	// RemovePod removes pod from the snapshot.
+	RemovePod(namespace string, podName string) error
+	// AddNodeWithPods adds a node and set of pods to be scheduled to this node to the snapshot.
+	AddNodeWithPods(node *apiv1.Node, pods []*apiv1.Pod) error
+
+	// GetAllPods returns list of all the pods in snapshot
+	GetAllPods() ([]*apiv1.Pod, error)
+	// GetAllNodes returns list of ll the nodes in snapshot
+	GetAllNodes() ([]*apiv1.Node, error)
+
+	// Fork creates a fork of snapshot state. All modifications can later be reverted to moment of forking via Revert()
+	// Forking already forked snapshot is not allowed and will result with an error.
+	Fork() error
```

Comment:
I understand why limiting to a single fork makes implementation easier, but it makes for a pretty poor interface - you suddenly need to worry about fork-safety of every method (this is especially painful for processors called in forked context).

Reply:
Agreed. But so far we did not find a use case for multiple forks. Therefor I would opt for keeping the limitation in interface.